File: effective-rust/java-http-server/src/main/java/com/effective/http/HttpServer.java
package com.effective.http;

import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

public class HttpServer {
    private static String body;

    public static void main(String[] args) throws IOException {
        int port = 8080;
        ServerSocket serverSocket = new ServerSocket(port);
        Executor executor = Executors.newFixedThreadPool(4);
        while (true) {
            final Socket clientSocket = serverSocket.accept();
            executor.execute(() -> {
                try {
                    BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
                    BufferedWriter out = new BufferedWriter(new OutputStreamWriter(clientSocket.getOutputStream()));

                    String request;
                    while ((request = in.readLine()) != null) {
                        if (request.isEmpty()) {
                            break;
                        }
                    }
                    if (body == null) {
                        InputStream bodyInputStream = HttpServer.class.getResourceAsStream("/index.html");
                        body = new BufferedReader(new InputStreamReader(bodyInputStream))
                                .lines().collect(Collectors.joining("\n"));
                    }
                    String response = String.format(
                            "HTTP/1.1 200 OK\r\nContent-Length: %d\r\n\r\n%s",
                            body.length(), body
                    );
                    out.write(response);
                    out.close();
                    in.close();
                    clientSocket.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            });
        }
    }
}

File: effective-kotlin/src/main/kotlin/com/effective/kotlin/coroutines/HttpServer.kt
package com.effective.kotlin.coroutines

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.*
import java.net.ServerSocket

const val RESPONSE_BODY = """
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Kotlin HTTP Server</title>
</head>
<body>
<center>
    <h1>Kotlin server is working!</h1>
</center>
</body>
</html>
"""

fun main() {
    val portNumber = 8080
    val socket = ServerSocket(portNumber)
    while (true) {
        val clientSocket = socket.accept()
        GlobalScope.launch {
            withContext(Dispatchers.IO) {
                try {
                    val input = BufferedReader(InputStreamReader(clientSocket.getInputStream()))
                    val output = BufferedWriter(OutputStreamWriter(clientSocket.getOutputStream()))
                    var request: String
                    while (input.readLine().also { request = it } != null) {
                        if (request.isEmpty()) {
                            break
                        }
                    }
                    val response = String.format(
                        "HTTP/1.1 200 OK\r\nContent-Length: %d\r\n\r\n%s",
                        RESPONSE_BODY.length, RESPONSE_BODY
                    )
                    output.write(response)
                    output.close()
                    input.close()
                    clientSocket.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }
}